File: shared/config/logger.py
"""
Configuração de logging estruturado para microsserviços Python
Usa structlog para logging estruturado e integração com Jaeger para distributed tracing
"""

import logging
import sys
import os
from typing import Any, Dict
import structlog
from structlog.stdlib import LoggerFactory
import uuid
from datetime import datetime


def configure_logging(service_name: str) -> structlog.BoundLogger:
    """
    Configura logging estruturado para o microsserviço
    
    Args:
        service_name: Nome do microsserviço
        
    Returns:
        Logger configurado
    """
    log_level = os.getenv('LOG_LEVEL', 'INFO').upper()
    log_format = os.getenv('LOG_FORMAT', 'json')
    
    # Configurar logging padrão do Python
    logging.basicConfig(
        format="%(message)s",
        stream=sys.stdout,
        level=getattr(logging, log_level)
    )
    
    # Processadores para formato JSON
    json_processors = [
        structlog.contextvars.merge_contextvars,
        structlog.stdlib.add_log_level,
        structlog.stdlib.add_logger_name,
        structlog.processors.TimeStamper(fmt="iso"),
        structlog.processors.StackInfoRenderer(),
        structlog.processors.format_exc_info,
        structlog.processors.UnicodeDecoder(),
        structlog.processors.JSONRenderer()
    ]
    
    # Processadores para formato console (desenvolvimento)
    console_processors = [
        structlog.contextvars.merge_contextvars,
        structlog.stdlib.add_log_level,
        structlog.stdlib.add_logger_name,
        structlog.processors.TimeStamper(fmt="%Y-%m-%d %H:%M:%S"),
        structlog.processors.StackInfoRenderer(),
        structlog.processors.format_exc_info,
        structlog.dev.ConsoleRenderer()
    ]
    
    processors = json_processors if log_format == 'json' else console_processors
    
    # Configurar structlog
    structlog.configure(
        processors=processors,
        wrapper_class=structlog.stdlib.BoundLogger,
        context_class=dict,
        logger_factory=LoggerFactory(),
        cache_logger_on_first_use=True,
    )
    
    # Criar logger com contexto do serviço
    logger = structlog.get_logger()
    logger = logger.bind(service=service_name)
    
    return logger


def generate_trace_id() -> str:
    """Gera um trace ID único"""
    return f"{int(datetime.now().timestamp() * 1000)}-{uuid.uuid4().hex[:9]}"


class RequestLoggingMiddleware:
    """
    Middleware FastAPI para logging de requisições
    """
    
    def __init__(self, logger: structlog.BoundLogger):
        self.logger = logger
    
    async def __call__(self, request, call_next):
        import time
        
        # Obter ou gerar trace ID
        trace_id = request.headers.get('x-trace-id', generate_trace_id())
        
        # Adicionar trace ID ao contexto
        structlog.contextvars.bind_contextvars(trace_id=trace_id)
        
        # Log da requisição
        self.logger.info(
            "incoming_request",
            method=request.method,
            path=request.url.path,
            query=str(request.query_params),
            client_ip=request.client.host if request.client else None,
            user_agent=request.headers.get('user-agent')
        )
        
        # Processar requisição
        start_time = time.time()
        response = await call_next(request)
        duration = (time.time() - start_time) * 1000  # em ms
        
        # Adicionar trace ID ao header da resposta
        response.headers['X-Trace-Id'] = trace_id
        
        # Log da resposta
        log_level = "error" if response.status_code >= 400 else "info"
        self.logger.log(
            log_level,
            "request_completed",
            method=request.method,
            path=request.url.path,
            status_code=response.status_code,
            duration_ms=f"{duration:.2f}"
        )
        
        # Limpar contexto
        structlog.contextvars.clear_contextvars()
        
        return response


def log_exception(logger: structlog.BoundLogger, exc: Exception, context: Dict[str, Any] = None):
    """
    Loga uma exceção com contexto adicional
    
    Args:
        logger: Logger configurado
        exc: Exceção a ser logada
        context: Contexto adicional (opcional)
    """
    log_data = {
        "exception_type": type(exc).__name__,
        "exception_message": str(exc),
        "exception_traceback": True
    }
    
    if context:
        log_data.update(context)
    
    logger.error("exception_occurred", **log_data, exc_info=exc)


# Exemplo de uso
if __name__ == "__main__":
    # Configurar logger
    logger = configure_logging("example-service")
    
    # Logs de exemplo
    logger.info("service_started", version="1.0.0")
    logger.debug("debug_message", data={"key": "value"})
    logger.warning("warning_message", reason="test")
    
    try:
        raise ValueError("Exemplo de erro")
    except Exception as e:
        log_exception(logger, e, {"additional_context": "test"})

File: shared/config/logger.js
// Configuração de logging estruturado para microsserviços Node.js
// Usa Winston para logging e integração com Jaeger para distributed tracing

const winston = require('winston');
const { format } = winston;

// Formato customizado para logs estruturados
const structuredFormat = format.combine(
  format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
  format.errors({ stack: true }),
  format.metadata({ fillExcept: ['message', 'level', 'timestamp', 'service'] }),
  format.json()
);

// Formato para console (desenvolvimento)
const consoleFormat = format.combine(
  format.colorize(),
  format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
  format.printf(({ timestamp, level, message, service, ...metadata }) => {
    let msg = `${timestamp} [${service}] ${level}: ${message}`;
    if (Object.keys(metadata).length > 0) {
      msg += ` ${JSON.stringify(metadata)}`;
    }
    return msg;
  })
);

/**
 * Cria um logger configurado para o microsserviço
 * @param {string} serviceName - Nome do microsserviço
 * @returns {winston.Logger} Logger configurado
 */
function createLogger(serviceName) {
  const logLevel = process.env.LOG_LEVEL || 'info';
  const logFormat = process.env.LOG_FORMAT || 'json';

  const transports = [
    new winston.transports.Console({
      format: logFormat === 'json' ? structuredFormat : consoleFormat
    })
  ];

  // Em produção, adicionar transporte para arquivo ou serviço externo
  if (process.env.NODE_ENV === 'production') {
    transports.push(
      new winston.transports.File({
        filename: `logs/${serviceName}-error.log`,
        level: 'error',
        format: structuredFormat
      }),
      new winston.transports.File({
        filename: `logs/${serviceName}-combined.log`,
        format: structuredFormat
      })
    );
  }

  return winston.createLogger({
    level: logLevel,
    defaultMeta: { service: serviceName },
    transports
  });
}

/**
 * Middleware Express para logging de requisições
 * @param {winston.Logger} logger - Logger do Winston
 */
function requestLogger(logger) {
  return (req, res, next) => {
    const startTime = Date.now();

    // Log da requisição
    logger.info('Incoming request', {
      method: req.method,
      path: req.path,
      query: req.query,
      ip: req.ip,
      userAgent: req.get('user-agent'),
      traceId: req.headers['x-trace-id']
    });

    // Interceptar o fim da resposta
    res.on('finish', () => {
      const duration = Date.now() - startTime;
      const logLevel = res.statusCode >= 400 ? 'error' : 'info';

      logger.log(logLevel, 'Request completed', {
        method: req.method,
        path: req.path,
        statusCode: res.statusCode,
        duration: `${duration}ms`,
        traceId: req.headers['x-trace-id']
      });
    });

    next();
  };
}

/**
 * Middleware Express para adicionar trace ID
 */
function traceIdMiddleware(req, res, next) {
  const traceId = req.headers['x-trace-id'] || generateTraceId();
  req.traceId = traceId;
  res.setHeader('X-Trace-Id', traceId);
  next();
}

/**
 * Gera um trace ID único
 */
function generateTraceId() {
  return `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
}

module.exports = {
  createLogger,
  requestLogger,
  traceIdMiddleware
};
